### PR TITLE
Sentry 2.0 [pull request #2]

### PR DESCRIPTION
**Pull request by nokitakaze**
* Set PHP 7.1 as **required** version because RC is already ready
* Raven Client initialization refactoring
* PHP 5.4: Force short syntax arrays (add rule to code style checking script)
* PHPDoc changes
* ~~Delete old PHP version~~

Issue #424

I'm afraid we must do coverage for all code we have.

### CODE REVIEW

**Comment by ste93cry** [comment]
What about changing the `.php_cs` file to configure short syntax arrays? This way linting will be automatic

**Comment by nokitakaze** [comment]
I think I should add something like this
```
 ->fixers(...blahblahblah...)
```
But I don't know how and I can't find solution [on the Net](http://lmgtfy.com/?q=%22phpcs%22+fixer+short+array). So if you would be so kind...

**Comment by ste93cry** [comment]
It should be

```php
->setRules(array(
    'array_syntax' => array('syntax' => 'short'),
)
```

**Comment by nokitakaze** [comment]
I'm afraid php-cs-fixer 1.8.0 does not contain setRules

**Comment by ste93cry** [comment]
We just need to update the PHP-CS-Fixer to a newer version

File: lib/Raven/Breadcrumbs/ErrorHandler.php [hunk 1]
```diff
@@ -41,7 +41,11 @@ public function handleError($code, $message, $file = '', $line = 0, $context = a
 
     public function install()
     {
-        $this->existingHandler = set_error_handler(array($this, 'handleError'), E_ALL);
+        $this->existingHandler = set_error_handler(
```
Comment (nokitakaze) [inline lib/Raven/Breadcrumbs/ErrorHandler.php]: > Why did you change this? 

Because we use PHP 5.5+ not a PHP 4 or something old shit like that

> we have one more function call for nothing

So what? 
https://3v4l.org/2RuH4 9 \* 10 ^ -8 second per a function call. 

If you want to save nanoseconds per function calls, just choose some other language which supports JIT. I.e. C# or Java or PHP 8

Comment (ste93cry) [inline lib/Raven/Breadcrumbs/ErrorHandler.php]: There is no valid reason to wrap a function call in a lambda as you're doing. PHP 5.5 or PHP 4 have nothing to do with this argument. A [callable](http://php.net/manual/en/language.types.callable.php) in PHP is represented in various forms, one of which is an array.

Comment (nokitakaze) [inline lib/Raven/Breadcrumbs/ErrorHandler.php]: Yes. More than that.
You can read files with
```php
$fh = fopen('/tmp/test.zip', 'r');
$data = '';
while(!feof($fh)) {
    $data .= fread($fh, 1024);
}
fclose($fh);
```

But I think you just use file_get_contents. And I think we want to use **modern** language. So why should we use some things which was introduced because of some guys who doesn\'t know how to  develop programming language?

Comment (ste93cry) [inline lib/Raven/Breadcrumbs/ErrorHandler.php]: As I said below, [callable](http://php.net/manual/en/language.types.callable.php) is not old. Lambda is not appropriate in that case because it introduces an overhead for no reason. One thing is to have to call a function (in that case `handleError`) which takes a different set of arguments than the input, so you can proxy the call through a lambda (this is just an example) and bind different arguments or adds more, one thing is call a lambda just to call another function that takes the same arguments and can be called with one line and the standard PHP syntax. I'm sorry, but what you're doing have no pros and a lot of cons

**Comment by nokitakaze** [comment]
@dcramer, we want to know your opinion

1. Minimal PHP version
2. Lambda instead of old syntax [$this, 'method_name']

**Comment by ste93cry** [comment]
>Lambda instead of old syntax [$this, 'method_name']

Again, old syntax as you call it is not old syntax. They serve different purposes! Said this, do whatever you want

**Comment by nokitakaze** [comment]
Ok. I can not wait anymore. I will change back closures

**Comment by ste93cry** [comment]
Before merging, you should also squash your commits

**Comment by nokitakaze** [comment]
default
I don\'t know how to do it. I have merged renews to this PR. So "combine" function is blocked

**Comment by nokitakaze** [comment]
I think I can just copy all my files and then paste them overwriting existed after merge 2.0, but I do not want to do it. 
I don\'t think it is necessary to have only one commit per pull request.

File: lib/Raven/Client.php [hunk 2]
```diff
@@ -683,7 +690,11 @@ protected function registerShutdownFunction()
     {
         if (!$this->_shutdown_function_has_been_set) {
             $this->_shutdown_function_has_been_set = true;
-            register_shutdown_function(array($this, 'onShutdown'));
```
Comment (nokitakaze) [inline lib/Raven/Client.php]: Erm...
https://github.com/nokitakaze/sentry-php/blob/2d338f555d2c8e771d17019ae881b572db0c35a1/lib/Raven/Client.php#L847

**Comment by ste93cry** [comment]
After the last merge commit some things still use the long array syntax, so if you could re-run the PHPCS you would fix it.

>I don't think it is necessary to have only one commit per pull request.

I never said this, but having 11 commits in a PR makes me wonder if there is something strange. Also, there is no need to know that you merged the changes of the branch this PR is based on with your own branch. Plus, You should _rebase_ your changes on top of 2.0 instead of _merging_ them. Finally, you have to consider that each single commit of the PRs will appear in the history of this repository, and it's really difficult to keep track of what happened if there are everytime something like 11 commits related to one change only

File: lib/Raven/Processor/SanitizeDataProcessor.php [hunk 3]
```diff
@@ -98,7 +98,11 @@ public function sanitizeHttp(&$data)
             }
         }
         if (!empty($http['data']) && is_array($http['data'])) {
-            array_walk_recursive($http['data'], array($this, 'sanitize'));
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeDataProcessor.php]: There is no need for another function call here (lambda). Please use the callable notation with array syntax

File: lib/Raven/Processor/SanitizeDataProcessor.php [hunk 4]
```diff
@@ -108,7 +112,11 @@ public function sanitizeStacktrace(&$data)
             if (empty($frame['vars'])) {
                 continue;
             }
-            array_walk_recursive($frame['vars'], array($this, 'sanitize'));
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeDataProcessor.php]: Same as above, there is no need for another function call here (lambda)

File: lib/Raven/Processor/SanitizeDataProcessor.php [hunk 5]
```diff
@@ -127,7 +135,11 @@ public function process(&$data)
             $this->sanitizeHttp($data);
         }
         if (!empty($data['extra'])) {
-            array_walk_recursive($data['extra'], array($this, 'sanitize'));
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeDataProcessor.php]: There is no need for another function call here (lambda)

File: lib/Raven/Processor/SanitizeHttpHeadersProcessor.php [hunk 6]
```diff
@@ -40,7 +40,10 @@ public function __construct(Client $client)
      */
     public function setProcessorOptions(array $options)
     {
-        $this->httpHeadersToSanitize = array_merge($this->getDefaultHeaders(), isset($options['sanitize_http_headers']) ? $options['sanitize_http_headers'] : array());
+        $this->httpHeadersToSanitize = array_merge(
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeHttpHeadersProcessor.php]: There is no need for another function call here (lambda)

**Comment by nokitakaze** [comment]
> 11 commits related to one change only

Short array syntax, phpdoc, deleting old php version and set 7.1 as required. And I want to keep commit because of how many things I changed in this pull request.

I have never done a rebase. I don\'t even clearly know how it works.

File: lib/Raven/Processor/SanitizeDataProcessor.php [hunk 7]
```diff
@@ -127,7 +135,11 @@ public function process(&$data)
...
+            array_walk_recursive(
+                $data['extra'], function (&$value, $key) {
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeDataProcessor.php]: There is no need for another function call here (lambda)

File: lib/Raven/Processor/SanitizeDataProcessor.php [hunk 8]
```diff
@@ -108,7 +112,11 @@ public function sanitizeStacktrace(&$data)
             if (empty($frame['vars'])) {
                 continue;
             }
-            array_walk_recursive($frame['vars'], array($this, 'sanitize'));
+            array_walk_recursive(
+                $frame['vars'], function (&$value, $key) {
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeDataProcessor.php]: Same as above, there is no need for another function call here (lambda)

File: lib/Raven/Processor/SanitizeDataProcessor.php [hunk 9]
```diff
@@ -98,7 +98,11 @@ public function sanitizeHttp(&$data)
             }
         }
         if (!empty($http['data']) && is_array($http['data'])) {
-            array_walk_recursive($http['data'], array($this, 'sanitize'));
+            array_walk_recursive(
+                $http['data'], function (&$value, $key) {
```
Comment (ste93cry) [inline lib/Raven/Processor/SanitizeDataProcessor.php]: There is no need for another function call here (lambda). Please use the callable notation with array syntax

**Comment by ste93cry** [comment]
>PHP 5.4: Short syntax arrays
>PHP 5.3: Force short array syntax (add rule to code style checking script)

For example, these two commits both change the same thing, and could be squashed together. There is no need to have one commit that applies summarily the short array syntax and another one that updates the PHPCS configuration and re-apply again the linting through all files finding many of them that somehow were not properly linted in the previous commit. This two commits will only generate confusion in the Git history. As I already said, I'm absolutely not saying that a PR should have just 1 commit or that a feature could not be delivered in more than 1 commit, I'm just saying that if we can squash together some commits that does the same thing then it's better

**Comment by nokitakaze** [comment]
I will combine all of them to one commit, but at this moment my working computer under heavy load. After calculation end I will delete all lambda I created and combine commits

File: lib/Raven/Client.php [hunk 10]
```diff
@@ -881,9 +1016,12 @@ public function capture($data, $stack = null, $vars = null)
             }
 
             if (!isset($data['stacktrace']) && !isset($data['exception'])) {
-                $data['stacktrace'] = array(
-                    'frames' => Stacktrace::fromBacktrace($this, $stack, isset($stack['file']) ? $stack['file'] : __FILE__, isset($stack['line']) ? $stack['line'] : __LINE__)->getFrames(),
-                );
+                $data['stacktrace'] = [
+                    'frames' => Stacktrace::fromBacktrace(
+                        $this, $stack, isset($stack['file']) ? $stack['file'] : __FILE__,
+                        isset($stack['line']) ? $stack['line'] : __LINE__
```
Comment (ste93cry) [inline lib/Raven/Client.php]: As `__LINE__` macro is used here to set in the stacktrace the line that generated the first frame, wouldn't be better to revert back these lines to just 1? This way the line will be the correct one (the one where the method is called)

Comment (nokitakaze) [inline lib/Raven/Client.php]: That is not my lines. I don\'t what they are doing. I just wrap these lines of code.

Comment (ste93cry) [inline lib/Raven/Client.php]: Please revert back the changes, so that the line of the first frame is set to the real line the method `fromBacktrace` has been called from.

Comment (nokitakaze) [inline lib/Raven/Client.php]: There is no need to show one particular line. The difference in one line does not make any changes when we are talking about debug. Otherwise, this is a very bad code.

Comment (ste93cry) [inline lib/Raven/Client.php]: In a debug backtrace the frame line corresponds to the line the method is called from, not one random line (even though there is formatting applied). With this change the user will see a wrong line

Comment (nokitakaze) [inline lib/Raven/Client.php]: If this changes breaks something why there is no tests to show it?

Comment (ste93cry) [inline lib/Raven/Client.php]: In a debug backtrace the frame line corresponds to the line the method is called from, not one random line (even though there is formatting applied). With this change the user will see a wrong line highlighted in Sentry's backend

Comment (nokitakaze) [inline lib/Raven/Client.php]: Maybe the good thing is to set
```
__LINE__ - 2
```
?

https://github.com/getsentry/sentry-php/blob/2.0/lib/Raven/Stacktrace.php#L78 I don\'t see anything beside debug information

Comment (ste93cry) [inline lib/Raven/Client.php]: >Maybe the good thing is to set

That would definitely work, although it seems a bit strange from a code perspective (at least in my opinion). I don't really see why this line should be splitted into multiple lines (also it makes no sense to me to move the first 3 arguments to one line and the other to another one), but if you really want to do that use the piece of code you posted to set the correct line in the stacktrace. Attached you can find some screenshots of the comparison between before and after your change

### Before
before

### After
after

### After change with `__LINE__ - 2`
after_line_subtraction

Now, although indeed there will be a lot of cases where something similar to the last screenshot happens in user code (and consequently we can't do anything to colorize the correct lines in Sentry) I'm not sure this change is worth. But these are just my two cents

Comment (nokitakaze) [inline lib/Raven/Client.php]: > should be splitted into multiple lines

The line is too long. Lines should not to be longer than 120 characters.

Comment (ste93cry) [inline lib/Raven/Client.php]: >Lines should not to be longer than 120 characters.

I absolutely agree on the topic, although 120 characters is maybe a common setting or something you decided yourself. By the way, I'm not interested in starting a discussion for a so small change, I don't think it's useful at all to discuss on every change you do, so if you evaluated all the pros/cons of this change and still think that it's worth although it "breaks" the colorization of the correct line in Sentry, use `__LINE__ - 2` and I will be ok with that

**Comment by ste93cry** [comment]
If you can split into another commit the changes about the client initialization refactoring from the code linting, then I'm ok with this PR and will approve it for merging

**Comment by nokitakaze** [comment]
Why are you telling me this just now? This is sense of humor or what? I already combined all commits to single commit. At 2:00 AM I'm not gonna fix it.

**Comment by ste93cry** [comment]
>I'm just saying that if we can squash together some commits that does the same thing then it's better

I never said to squash everything into one commit. I repeated so many times I can't count them that I'm not against having a single commit per PR, but that commits that does the same thing should be squashed together (and this is anyway my opinion). You understood to squash everything into one commit...

**Comment by nokitakaze** [comment]
There was no way to combine only specific ones. Or I had to start writing the whole pull request from scratch, having spent a couple of hours on it

**Comment by ste93cry** [comment]
>There was no way to combine only specific ones. Or I had to start writing the whole pull request from scratch

This is not true at all, btw I had to force-push to your branch because I rewritten the Git history using the rebase command to split the commits. Everything seems fine and tests still pass, so PR is good for me 👍

**Comment by nokitakaze** [comment]
David did not tell us minimal version :(

Okay. At this moment it will be PHP 5.5
